### PR TITLE
Further improvements of guide.rst

### PR DESCRIPTION
**Pull request by sylee957**
#### Brief description of what is fixed or changed

There were some python 2 printing commands, and the tutorial had some more errors that have not been addressed in #18100. 

I've removed some bad practices like overriding `is_finite = True`, and removed `_eval_is_zero` there because it would not likely be testable from the logic there without introducing the concept of an unevaluated object, which may not be suitable for the introductory tutorial.

The only things I haven't touched are some shell commands from ondrej's approach section, which may not be doctested, and code snippets there can also be obsolete, but this may not be much of a concern.

I also don't think that the class structure of basic is accurate there, but haven't updated that part yet.

#### Release Notes

NO ENTRY

### CODE REVIEW

File: doc/src/guide.rst [hunk 1]
```diff
@@ -279,52 +284,86 @@ Functions
...
-            if arg.is_negative:
-                return S.NegativeOne
-            if isinstance(arg, Basic.Mul):
-                coeff, terms = arg.as_coeff_mul()
-                if not isinstance(coeff, Basic.One):
-                    return cls(coeff) * cls(Basic.Mul(*terms))
+The applied function ``sign(x)`` is constructed using
 
-        is_finite = True
+    >>> sign(x)
```
Comment (ShubhamKJha) [inline doc/src/guide.rst]: Maybe, a
```
from sympy.abc import x
```

File: doc/src/guide.rst [hunk 2]
```diff
@@ -279,52 +284,86 @@ Functions
...
 
-and that's it. The ``_eval_*`` functions are called when something is needed.
 The ``eval`` is called when the class is about to be instantiated and it
-should return either some simplified instance of some other class or if the
-class should be unmodified, return ``None`` (see ``core/function.py`` in
-``Function.__new__`` for implementation details). See also tests in
-`sympy/functions/elementary/tests/test_interface.py
-<https://github.com/sympy/sympy/blob/master/sympy/functions/elementary/tests/test_interface.py>`_ that test this interface. You can use them to create your own new functions.
+should return either an another SymPy object or ``None``.
+If it is returned ``None``, it becomes a
```
Comment (ShubhamKJha) [inline doc/src/guide.rst]: Is the sentence complete here?

File: doc/src/guide.rst [hunk 3]
```diff
@@ -279,52 +284,86 @@ Functions
...
+(see ``core/function.py`` in ``Function.__new__`` for implementation details)
 
-The applied function ``sign(x)`` is constructed using
-::
+Here are some examples that how ``eval`` works
+
```
Comment (ShubhamKJha) [inline doc/src/guide.rst]: I think `::` is added to create a block in `rst` document. Should be done to separate the code block here.

File: doc/src/guide.rst [hunk 4]
```diff
@@ -279,52 +284,86 @@ Functions
...
-    sign
+    >>> x = Symbol('x', negative=True)
+    >>> sign(x)
+    -1
+
+The ``_eval_*`` functions are automatically called when something is
+needed.
+For example, ``conjugate`` or ``is_finite`` will automatically call
+``_eval_is_finite``, ``_eval_conjugate`` if you have these defined.
+
```
Comment (ShubhamKJha) [inline doc/src/guide.rst]: `::` should be added here as well.

File: doc/src/guide.rst [hunk 5]
```diff
@@ -279,52 +284,86 @@ Functions
...
+    >>> class sign(Function):
+    ...     nargs = 1
+    ...
+    ...     @classmethod
+    ...     def eval(cls, arg):
+    ...         if arg is S.NaN:
+    ...             return S.NaN
+    ...         if arg is S.Zero:
+    ...             return S.Zero
+    ...         if arg.is_positive:
```
Comment (ShubhamKJha) [inline doc/src/guide.rst]: Maybe it should be updated to `arg.is_extended_positive`.

**Comment by asmeurer** [comment]
I still think we should delete this thing https://github.com/sympy/sympy/pull/2807

**Comment by sylee957** [comment]
I'd only suggest that merging this can put out some fire, even if some discussions are lagging down merging #2807, and other PRs would just need a rebase.

**Comment by oscarbenjamin** [comment]
I'll merge this for now but I agree that this guide should be replaced.

**Pull request merged**